File: src/main/typescript/org/bubblecloud/webvr/Network.ts
import {Envelope} from "./model/Envelope";
import {Message} from "./model/Message";
export class Network {

    webSocketUrl : string = "ws://" + window.location.hostname+(location.port ? ':' + location.port : "") + "/ws";
    socket : WebSocket;

    constructor() {

    }

    connect = () => {

        this.socket = new WebSocket(this.webSocketUrl);

        this.socket.onopen = () => {
            console.log("Socket opened: " + this.socket.url);

            var handshakeRequest = new Message();
            handshakeRequest.type = "handshake-request";
            handshakeRequest.properties = {
                "software" : "kotlin-web-vr",
                "protocol-dialect": "vr-state-synchronisation",
                "protocol-versions" : ["0.9", "1.0"]
            };

            var envelope = new Envelope();
            envelope.messages = [handshakeRequest];

            this.send(envelope);
        };

        this.socket.onclose = () =>  {
            console.log("Socket closed: " + this.socket.url);
        };

        this.socket.onerror = (event: Event) => {
            console.log("Socket error: " + this.socket.url + " - " + event);
        };

        this.socket.onmessage = (messageEvent: MessageEvent) => {
            this.onMessage(messageEvent);
        }

    };

    private onMessage = (messageEvent: MessageEvent) => {
        this.receive(JSON.parse(messageEvent.data))
    }

    receive = (envelope : Envelope) => {
        console.log("Received: " + JSON.stringify(envelope));
    }

    send = (envelope : Envelope) => {
        this.socket.send(JSON.stringify(envelope));
    }

}

File: server/src/main/kotlin/org/bubblecloud/webvr/Network.kt
package org.bubblecloud.webvr

import logger
import org.bubblecloud.webvr.model.Envelope
import org.bubblecloud.webvr.model.Message
import org.bubblecloud.webvr.model.Node
import org.bubblecloud.webvr.model.Session
import org.bubblecloud.webvr.util.Mapper
import org.glassfish.grizzly.websockets.Broadcaster
import org.glassfish.grizzly.websockets.OptimizedBroadcaster
import org.glassfish.grizzly.websockets.WebSocket
import java.util.*
import java.util.logging.Level

class Network() {

    private val log = logger()

    private var broadcaster: Broadcaster = OptimizedBroadcaster()

    private val mapper = Mapper()

    private val sessions: MutableMap<WebSocket, Session> = HashMap()

    @Synchronized fun addSession(session: Session) {
        log.info("Session connected: " + session.remoteHost + ":" + session.remotePort)
        sessions[session.socket] = session
    }

    @Synchronized fun removeSession(socket: WebSocket) {
        val session: Session? = sessions[socket]
        if (session != null) {
            log.info("Session disconnected: ${session.remoteHost}:${session.remotePort}")
            sessions.remove(socket)
        }
    }

    @Synchronized fun receive(socket: WebSocket, jsonString: String) {
        try {
            val session = sessions[socket] ?: return
            val envelope = mapper.readValue(jsonString, Envelope::class.java)

            val values = mapper.readValuesFromEnvelope(envelope)

            val nodes : MutableList<Any> = mutableListOf()

            for (value in values) {
                if (value is Message) {
                    if (value.type.equals("handshake-request")) {
                        log.info("Handshake accepted : ${session.remoteHost}:${session.remotePort} (${value.properties["software"]})")

                        val responseEnvelope = Envelope()
                        val handshakeResponse = Message("handshake-response", mapOf(
                                "software" to "kotlin-web-vr",
                                "protocol-dialect" to "vr-state-synchronisation",
                                "protocol-version" to "1.0",
                                "accepted" to "true"
                        ))
                        val values : MutableList<Any> = mutableListOf()
                        values.add(handshakeResponse)
                        values.addAll(CELL.getNodes())
                        mapper.writeValuesToEnvelope(responseEnvelope, values)
                        socket.send(mapper.writeValue(responseEnvelope))
                    }
                }
                if (value is Node) {
                    if (!value.removed) {
                        if (CELL.hasNode(value.uri)) {
                            CELL.updateNode(value)
                        } else {
                            CELL.addNode(value)
                        }
                    } else {
                        CELL.removeNode(value.uri)
                    }
                    nodes.add(value)
                }
            }

            if (nodes.size > 0) {
                val broadcastEnvelope = Envelope()
                mapper.writeValuesToEnvelope(broadcastEnvelope, nodes)
                broadcast(broadcastEnvelope)
            }

        } catch (ex: Exception) {
            log.log(Level.SEVERE, "Failed to process incoming message.", ex)
        }
    }

    @Synchronized fun broadcast(envelope: Envelope) {
        val jsonString = mapper.writeValue(envelope)
        broadcaster.broadcast(sessions.keys, jsonString)
    }

    fun  getSession(socket: WebSocket?): Session? {
        return sessions[socket]
    }


}